unittests/etl/TransformerTest.cpp

```diff
@@ -0,0 +1,151 @@
+    auto const response = std::make_optional<FakeFetchResponse>(blob);
+
+    ON_CALL(dataPipe_, popNext).WillByDefault([this, &response](auto) -> std::optional<FakeFetchResponse> {
+        if (state_.isStopping)
+            return std::nullopt;
+        return response;
+    });
+    ON_CALL(*rawBackendPtr, doFinishWrites).WillByDefault(Return(true));
+
+    // TODO: most of this should be hidden in a smaller entity that is injected into the transformer thread
+    EXPECT_CALL(dataPipe_, popNext).Times(AtLeast(1));
+    EXPECT_CALL(*rawBackendPtr, startWrites).Times(AtLeast(1));
+    EXPECT_CALL(*rawBackendPtr, writeLedger(_, _)).Times(AtLeast(1));
+    EXPECT_CALL(ledgerLoader_, insertTransactions).Times(AtLeast(1));
+    EXPECT_CALL(*rawBackendPtr, writeAccountTransactions).Times(AtLeast(1));
+    EXPECT_CALL(*rawBackendPtr, writeNFTs).Times(AtLeast(1));
+    EXPECT_CALL(*rawBackendPtr, writeNFTTransactions).Times(AtLeast(1));
+    EXPECT_CALL(*rawBackendPtr, doFinishWrites).Times(AtLeast(1));
+    EXPECT_CALL(ledgerPublisher_, publish(_)).Times(AtLeast(1));
+
+    transformer_ =
+        std::make_unique<TransformerType>(dataPipe_, mockBackendPtr, ledgerLoader_, ledgerPublisher_, 0, state_);
+
+    // after 10ms we start spitting out empty responses which means the extractor is finishing up
+    // this is normally combined with stopping the entire thing by setting the isStopping flag.
+    std::this_thread::sleep_for(std::chrono::milliseconds{10});
+    state_.isStopping = true;
+}
+
+TEST_F(ETLTransformerTest, DoesNotPublishIfCanNotBuildNextLedger)
+{
+    MockBackend* rawBackendPtr = static_cast<MockBackend*>(mockBackendPtr.get());
+    mockBackendPtr->cache().setFull();  // to avoid throwing exception in updateCache
+
+    auto const blob = hexStringToBinaryString(RAW_HEADER);
+    auto const response = std::make_optional<FakeFetchResponse>(blob);
+
+    ON_CALL(dataPipe_, popNext).WillByDefault(Return(response));
+    ON_CALL(*rawBackendPtr, doFinishWrites).WillByDefault(Return(false));  // emulate write failure
+
+    // TODO: most of this should be hidden in a smaller entity that is injected into the transformer thread
+    EXPECT_CALL(dataPipe_, popNext).Times(AtLeast(1));
+    EXPECT_CALL(*rawBackendPtr, startWrites).Times(AtLeast(1));
+    EXPECT_CALL(*rawBackendPtr, writeLedger(_, _)).Times(AtLeast(1));
+    EXPECT_CALL(ledgerLoader_, insertTransactions).Times(AtLeast(1));
+    EXPECT_CALL(*rawBackendPtr, writeAccountTransactions).Times(AtLeast(1));
+    EXPECT_CALL(*rawBackendPtr, writeNFTs).Times(AtLeast(1));
+    EXPECT_CALL(*rawBackendPtr, writeNFTTransactions).Times(AtLeast(1));
+    EXPECT_CALL(*rawBackendPtr, doFinishWrites).Times(AtLeast(1));
+
+    // should not call publish
+    EXPECT_CALL(ledgerPublisher_, publish(_)).Times(0);
+
+    transformer_ =
+        std::make_unique<TransformerType>(dataPipe_, mockBackendPtr, ledgerLoader_, ledgerPublisher_, 0, state_);
```

Comment:
There is no sleep and explicit waiting . I am puzzled why EXPECT_CALL still pass.

Reply:
Both transformer and extractor have `join` in the destructor :)